src/base/emu-i386/simx86/interp.c

```diff
@@ -393,6 +393,7 @@ static unsigned int _Interp86(unsigned int PC, int mod0);
 unsigned int Interp86(unsigned int PC, int mod0)
 {
     unsigned int ret = _Interp86(PC, mod0);
+    register_sim_pagefault_handler(NULL);
```

Comment:
But I meant not a global call-back,
but rather to pass the call-back as
an arg... This isn't pretty, yes, but
what I find absolutely impossible,
is that this API won't be available
for other users part of the time
(namely, until the above NULL
callback is "set"). Is this really a
good design for an API? Yes, one
can argue thread-safety is not a
problem here, but even with one
thread you can only afford a single
user, because if someone else
installs his call-back, the prev one
is overwritten.
So I suggested to pass a call-back
as an argument: at least, that makes
an API more or less usable.

Yes, it _is_ like the signal handling.
With signal handling you only have
one caller of your handler: the kernel.
But write_byte() and stuff do not look
like they are going to have only one
caller. If they are, then why, why just
to move them to the public place?

Reply:
I.e. `q = read_byte(addr, sim_exception_cb);`.
Then you can macrosify it to hide an extra arg...
It will at least work.
But I don't see why, why is this all?

Reply:
Doing it like sighandling, will mean
that you need to pass an identifications.
In sighandlers we have signum and siginfo,
so we can identify the caller (or, more
precisely, the reason for the call).
So you would still need to pass in some
ID to make it signal-alike, and that ID
should better be the call-back pointer.
(and of course, much better is to just
revert the patch that exports those)

Reply:
I must admit I do not quite understand what you mean here, especially this "is that this API won't be available for other users part of the time (namely, until the above NULL
callback is "set")". The default is that NULL callback, ie. if cpuemu is not used it will just call leavedos. All that I wanted with this change is to make sure that the longjmp is not used outside Interp86, where it makes no sense, and instead do something reasonable there.

Other users can also set a handler, but in the case of e.g. dosemu2 internally calling write_byte to make sure it works with both VGA and normal RAM, any access to undefined or read-only memory is fatal, and instead of letting the MMU crash dosemu this will be handled by software (with the advantage that write_byte cannot accidentally write to r/w memory that is not specifically mapped for DOS).

"and of course, much better is to just
revert the patch that exports those"
which patch do you mean here?

Passing an extra argument to read_byte is quite expensive... the other alternative I can think of (to avoid overwriting) is to have a single handler that calls emu_handle_pagefault only when Interp86 is active (which is like it is without this patch, except it would need a check like if in_interp86.., which is then similar to having a jmp_env pointer null or non-null) ... then you are back to square 1 but it is not any different to what dosemu_fault and e_emu_pagefault (which checks for InCompiledCode) are doing already.

Reply:
> I must admit I do not quite understand what you mean here

1. What will happen if you set the call-back,
and someone else is calling that API? Crash,
because longjmp() will be called from wrong
stack.
2. What will happen if someone else set his
call-back, and you overwritten it, and then
reset to NULL w/o restoring? Nothing good.

This just isn't the good API.

> Other users can also set a handler

How? You overwrite it w/o restoring.

> letting the MMU crash dosemu this will be handled by software

It is handled by software - in SIGSEGV
handler we dump stack traces and all
other things, even running gdb automatically.
And for fatal errors its perfectly fine.
You have only one place where this is
both not fatal AND signal-unwanting.
So why not to have this place use it
on its own? In the ideal world you could
just add the xx_safe() versions of the
API. But in this particular case you have
that nasty longjump that just doesn't let
you call this from anything else.

I really don't see why this API can't be
private, and another one, similar - public,
but w/o any longjmp?
Well, if you don't want 2 APIs for whatever
reason, maybe the compromise can be
to use the `(uint64_t)-1` return value as
an error indication? But I still do not
understand why would you want to replace
our beloved SIGSEGV+gdb with something
else...

> e_emu_pagefault (which checks for InCompiledCode) are doing already.

e_emu_pagefault() indeed checks InCompiledCode,
but e_emu_pagefault() is a part of simx86, as
well as InCompiledCode. I just now moved that
var out of the public header, and everything still
compiles - no one accesses it outside of simx86.
And similarly, you shouldn't use the jmpbuf that
is specific to simx86, outside of it. But with jmpbuf
it is much worse, as that will actually crash.

Reply:
OK, suppose we added the threaded JumpGen... :)
The speculative threads are not permitted
to write anywhere, but they should not crash
on write - just get an error. But main thread
is still allowed to read and write, so obviously
they'd want the separate handlers each.
The API that you propose, in inherently
thread-unsafe, and will just crash in that
scenario. Making that API private, and
even then either pass a local call-back
or use weird return values like `(uint64_t)-1`
will at least be thread-safe.

Note that when the API is private, you are
not bound by any "beauty" criteria. So if
it is private, you can just return the aggregate.
And IMHO that would be the best choice:
make it private, return aggregate, get rid
of any longjmp - thread-safety.
If there is a need for a public version, it
can be introduced too, but call leavedos()
instead of returning aggregate.

Reply:
Btw, with aggregate you can also return
the synthetic exception error code, which
likely makes sense. :)

Reply:
I think the problem you are trying to
solve, is, if you are to introduce both
versions of such API, one returning
the aggregate and another - calling
leavedos(), then how to share the
code between them, rather than to
do the full copy/paste.
But IMHO this is a minor problem,
to start with. Firstly, the public version
is hardly ever needed, and secondly...
well, it can still be done with macros.

```
#define do_read_byte(addr, exc_code)
...
uint8_t read_byte(int addr) { do_read_byte(addr, { leavedos(); }); }
struct my_aggregate read_byte_simx86(int addr) { do_read_byte(addr, { return aggregate; }); }
```

So, even if you need it public, and want
to share the code with private, and all that,
it is still possible to do in a reliable and
thread-safe manner, but with macros.
But we already have LOWMEM_READ_xx()
public API, and if the mapbase branch
works (it was very close to, the last time
I checked), that LOWMEM API will actually
became capable of accessing everything,
so then why something else public...

Reply:
What happens is that on entry of function X (in this case Interp86) it's set, and on exit it needs to be unset (back to default). If another place wants to use it it will need to do the same thing -- I was just assuming so far that only unrelated functions would set the handler, not e.g. a function called by Interp86.  This could then be improved by returning the old handler in the handler-setting function (like signal() and sigaction() do) so the old handler can be saved and restored upon function entry and exit, that is, a stack of handlers. That's still not thread safe but would solve your problems 1 and 2.

> we dump stack traces and all
> other things, even running gdb automatically.
well yes that is what `dosemu_error()` does in the default handler, it's exactly the same code that calls gdb already :)

But I think what you mean is to just have equivalents of all that code in dos2linux.c to a place under simx86, prefix with emu_ and use those (emu_read_byte() and co?) instead. emu_read_byte can still call generic helper functions in dos2linux.c then.